### DISCUSSION TITLE
Inconsistency in parameter type in virtual method.

### DISCUSSION THREAD

**Discussion by GianptDev** [godotengine/godot-proposals#6679] [Engine Core]
Hello, I'm using `godot.3.5.2.stable`
I'm doing a project and using the `StaticBody2d` node with a script, I'm overriding the `_input_event` and since I like type typing I've set the type of each paramenters like this:
```swift
func _input_event(viewport: Viewport, event: InputEvent, shape_idx: int) -> void:
```
but the engine gave me the error:
```
Parser Error: The function signature doesn't match the parent. Parent signature is: "void _input_event(Object, InputEvent, int)".
```
Looking at `CollisionObject2D`, the class that implement the method, the parameter of `viewport` is of type `Object` instead of `Viewport`, despite the parameter will always contain a viewport.

**Should this be considered a bug? It was intended to be like this?**

**Comment by AThousandShips** [comment]
It's defined as accepting an object, regardless of the name, might be an oversight as the signal has type set

Changing it would break compatibility which I doubt will happen in 3.x

**Reply by GianptDev** [reply]
But why it does not use `Viewport`? The variable contain a viewport.
This is a problem for the autocomplete in the editor because it does not get the properties and functions of the viewport but only of the object class.
It can be fixed and ignored doing a cast like this, but I says this is just unnecessary.
```swift
var realViewport: = viewport as Viewport;
```
There is any reason why it was chosen to use `Object` as the type?
_Maybe is a fix for an older version of the engine that could be removed in the current version._

**Reply by AThousandShips** [reply]
Probably an oversight, but it would break compatibility to fix it and it doesn't cause anything more than a bother with autocomplete so I don't think it's likely to change

Also it would likely only be restricted to `Node` as that is what the calling function argument is
